<a href="https://colab.research.google.com/github/prthameshhh/Quantitative-Analysis/blob/main/Quantitative_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Selected Top Performing 4 Sectors

- **FMCG**
- **Banking**
- **Healthcare**
- **Energy**


In [31]:
import yfinance as yf
import pandas as pd
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots
import plotly.io as pio
pio.templates.default = "plotly_white"

Downloading the price data for the stocks

In [32]:
# List of stock tickers
stock_tickers = ['RELIANCE.NS', 'SUNPHARMA.NS', 'ITC.NS', 'ADANIENT.NS']

# Define the start and end date
end_date = pd.Timestamp.today()
start_date = end_date - pd.DateOffset(months=3)

# Fetch stock data
stock_data = {}
for ticker in stock_tickers:
    stock_data[ticker] = yf.download(ticker, start=start_date, end=end_date)

# Combine the data into a single DataFrame in sequence
stocks_data = pd.DataFrame()
for ticker in stock_tickers:
    data = stock_data[ticker]
    data['Ticker'] = ticker
    stocks_data = pd.concat([stocks_data, data])

# Reset index for stocks_data
stocks_data.reset_index(inplace=True)

[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed


In [33]:
# Display combined data
stocks_data['Ticker'].value_counts()

Ticker
RELIANCE.NS     57
SUNPHARMA.NS    57
ITC.NS          57
ADANIENT.NS     57
Name: count, dtype: int64

In [34]:
# Display the first few rows of the dataset
print(stocks_data.head())

        Date         Open         High          Low        Close    Adj Close  \
0 2024-02-22  2936.300049  2969.899902  2916.000000  2963.500000  2963.500000   
1 2024-02-23  2979.000000  2995.100098  2966.699951  2987.250000  2987.250000   
2 2024-02-26  2987.100098  2989.050049  2965.000000  2974.649902  2974.649902   
3 2024-02-27  2966.050049  2999.899902  2956.100098  2971.300049  2971.300049   
4 2024-02-28  2966.000000  2982.550049  2900.350098  2911.250000  2911.250000   

    Volume       Ticker  
0  9246864  RELIANCE.NS  
1  7219292  RELIANCE.NS  
2  3756553  RELIANCE.NS  
3  5413022  RELIANCE.NS  
4  4323975  RELIANCE.NS  


Descriptive Statistics
Descriptive Statistics will provide summary statistics for each stock in the dataset. We’ll look at measures such as mean, median, standard deviation, and more for the Close prices:

In [35]:
# Descriptive Statistics for each stock
descriptive_stats = stocks_data.groupby('Ticker')['Close'].describe()
print(descriptive_stats)

              count         mean         std          min          25%  \
Ticker                                                                   
ADANIENT.NS    57.0  3115.355259  137.447100  2766.649902  3049.750000   
ITC.NS         57.0   424.231580   10.393417   404.450012   415.700012   
RELIANCE.NS    57.0  2913.844731   56.346406  2788.250000  2871.399902   
SUNPHARMA.NS   57.0  1557.097369   39.068311  1484.650024  1525.050049   

                      50%          75%          max  
Ticker                                               
ADANIENT.NS   3111.750000  3224.800049  3327.750000  
ITC.NS         425.899994   431.450012   441.000000  
RELIANCE.NS   2925.850098  2959.149902  3014.800049  
SUNPHARMA.NS  1552.099976  1593.599976  1629.250000  


The summary suggests a mix of investment opportunities, from high-growth and volatile stocks like ADANIENT.NS and RELIANCE.NS to more stable options like ITC.NS and SUNPHARMA.NS. we should align choices with risk tolerance, investment goals, and market outlook. Diversifying across these stocks could also be a strategy to balance risk and reward.

In [36]:
import pandas as pd
from plotly.subplots import make_subplots
import plotly.graph_objects as go

# Assuming your 'stocks_data' DataFrame is already loaded

# Convert 'Date' column to datetime format
stocks_data['Date'] = pd.to_datetime(stocks_data['Date'])

# Get unique tickers
tickers = pivot_data.columns

# Create a 2x2 subplot grid
rows, cols = 2, 2  # Adjust rows and cols based on number of tickers
fig = make_subplots(rows=rows, cols=cols, subplot_titles=tickers[:rows*cols])

# Iterate through each ticker and create traces
row, col = 1, 1
for ticker in tickers:
    fig.add_trace(
        go.Scatter(x=pivot_data.index, y=pivot_data[ticker], name=ticker),
        row=row, col=col
    )
    col += 1
    if col > cols:  # Move to next row after reaching the end of columns
        row += 1
        col = 1

# Update layout (adjust title as needed)
fig.update_layout(
    title_text='Time Series of Closing Prices for Each Stock',
    xaxis_title='Date',
    yaxis_title='Closing Price',
    showlegend=False  # Legend not required for separate subplots
)

# Adjust spacing between subplots (optional)
fig.update_layout(margin={'t': 50})  # Adjust top margin for subplot titles

# Show the plot
fig.show()


**Interpretation:**
Based on the visualizations, here are the potential trends for each stock:

- Adani Enterprises: Appears to be in an uptrend.
- ITC Ltd: Appears to be in a slightly uptrend.
- Reliance Industries: Appears to be in a downtrend.
- Sun Pharmaceutical Industries: Appears to be in a downtrend.


**Comparative Analysis:**

When comparing the trends of the four stocks, we observe distinct patterns:

- **Adani Enterprises**: This stock shows a clear uptrend, indicating strong positive momentum and potential for growth. Investors might find this stock appealing for long-term investment, given its upward trajectory.
  
- **ITC Ltd**: Although this stock is also in an uptrend, the trend is only slight. This suggests moderate growth potential. While it might not be as aggressive as Adani Enterprises, it still shows promise for gradual appreciation in value.
  
- **Reliance Industries**: This stock is in a downtrend, signaling potential challenges or negative sentiment in the market. Investors might approach this stock with caution, as it could indicate declining performance or other underlying issues.
  
- **Sun Pharmaceutical Industries**: Similar to Reliance Industries, this stock is experiencing a downtrend. This suggests a period of decline, and investors might need to consider the factors contributing to this trend before making any investment decisions.

In summary, Adani Enterprises and ITC Ltd show positive trends with varying degrees of optimism, while Reliance Industries and Sun Pharmaceutical Industries are currently facing downward trends, indicating potential concerns.


Volatility Analysis
Next, let’s focus on Volatility Analysis. We’ll calculate and compare the volatility (standard deviation) of the closing prices for each stock. It will give us an insight into how much the stock prices fluctuated over the period:



In [37]:
# Calculate the standard deviation (volatility) of closing prices for each stock
volatility = pivot_data.std()

# Print the standard deviation values
print(volatility)

Ticker
ADANIENT.NS     137.447100
ITC.NS           10.393417
RELIANCE.NS      56.346406
SUNPHARMA.NS     39.068311
dtype: float64


In [38]:
 # Volatility Analysis
volatility = pivot_data.std().sort_values(ascending=False)

fig = px.bar(volatility,
             x=volatility.index,
             y=volatility.values,
             labels={'y': 'Standard Deviation', 'x': 'Ticker'},
             title='Volatility of Closing Prices (Standard Deviation)')

# Show the figure
fig.show()

The bar chart and the accompanying data show the volatility (measured as standard deviation) of the closing prices for each stock. Here’s how they rank in terms of volatility:

**Volatility Analysis Interpretation:**

Based on the standard deviation of closing prices:

- **Adani Enterprises (ADANIENT.NS)**: Highest volatility (137.45). Significant price fluctuations, indicating higher risk and potential returns.
- **Reliance Industries (RELIANCE.NS)**: High volatility (56.35). Substantial price swings, implying considerable risk and potential rewards.
- **Sun Pharmaceutical Industries (SUNPHARMA.NS)**: Moderate volatility (39.07). Balanced risk-reward profile.
- **ITC Ltd (ITC.NS)**: Lowest volatility (10.39). Stable price movements, suitable for conservative investors seeking lower risk.

In summary, Adani Enterprises and Reliance Industries are higher risk with potential for higher returns, Sun Pharmaceutical Industries offers a moderate risk-reward balance, and ITC Ltd provides stability with lower risk.


Correlation Analysis
Next, we’ll perform a Correlation Analysis to understand how the stock prices of these companies are related to each other:

In [39]:
#Correlation Analysis
correlation_matrix = pivot_data.corr()

fig = go.Figure(data=go.Heatmap(
                    z=correlation_matrix,
                    x=correlation_matrix.columns,
                    y=correlation_matrix.columns,
                    colorscale='blues',
                    colorbar=dict(title='Correlation'),
                    ))

# Update layout
fig.update_layout(
    title='Correlation Matrix of Closing Prices',
    xaxis_title='Ticker',
    yaxis_title='Ticker'
)

# Show the figure
fig.show()

**Correlation Analysis Interpretation:**

The correlation coefficients between SUNPHARMA.NS and the other three stocks are all positive. This indicates that the closing price of SUNPHARMA.NS tends to move in the same direction as the closing prices of the other three stocks. Among these correlations, the coefficient between SUNPHARMA.NS and RELIANCE.NS is the highest at 0.8. This implies that the closing prices of SUNPHARMA.NS and RELIANCE.NS exhibit the closest movement out of the four stocks.

On the other hand, the correlation coefficient between ADANIENTINS.NS and ITC.NS is -0.2. This suggests a weak negative relationship between the two stocks, meaning that the closing price of ADANIENTINS.NS tends to move in the opposite direction of the closing price of ITC.NS.



Comparative Analysis
Now, let’s move on to Comparative Analysis. In this step, we’ll compare the performance of different stocks based on their returns over the period. We’ll calculate the percentage change in closing prices from the start to the end of the period for each stock:

In [40]:
# Calculating the percentage change in closing prices
percentage_change = ((pivot_data.iloc[-1] - pivot_data.iloc[0]) / pivot_data.iloc[0]) * 100

fig = px.bar(percentage_change,
             x=percentage_change.index,
             y=percentage_change.values,
             labels={'y': 'Percentage Change (%)', 'x': 'Ticker'},
             title='Percentage Change in Closing Prices')

# Show the plot
fig.show()

**Comparative Analysis of Overall Percentage Return:**

The overall percentage return for each stock over the specified period is as follows:

- **Adani Enterprises (ADANIENT.NS)**: The overall return is approximately -4.46%. This indicates a decrease in value over the period, suggesting a loss for investors who held this stock throughout the entire period.

- **ITC Ltd (ITC.NS)**: The overall return is approximately 4.91%. This indicates an increase in value over the period, suggesting a gain for investors who held this stock throughout the entire period.

- **Reliance Industries (RELIANCE.NS)**: The overall return is approximately -3.08%. This indicates a decrease in value over the period, suggesting a loss for investors who held this stock throughout the entire period.

- **Sun Pharmaceutical Industries (SUNPHARMA.NS)**: The overall return is approximately -1.10%. This indicates a decrease in value over the period, suggesting a loss for investors who held this stock throughout the entire period.

In summary, ITC Ltd is the only stock with a positive overall return, indicating a gain for investors. Adani Enterprises, Reliance Industries, and Sun Pharmaceutical Industries all experienced negative overall returns, resulting in losses for investors over the specified period.


Daily Risk Vs. Return Analysis
To perform a Risk vs. Return Analysis, we will calculate the average daily return and the standard deviation of daily returns for each stock. The standard deviation will serve as a proxy for risk, while the average daily return represents the expected return.


We will then plot these values to visually assess the risk-return profile of each stock. Stocks with higher average returns and lower risk (standard deviation) are generally more desirable, but investment decisions often depend on the investor’s risk tolerance: **bold text**

In [41]:
daily_returns = pivot_data.pct_change().dropna()

# Recalculating average daily return and standard deviation (risk)
avg_daily_return = daily_returns.mean()
risk = daily_returns.std()

# Creating a DataFrame for plotting
risk_return_df = pd.DataFrame({'Risk': risk, 'Average Daily Return': avg_daily_return})

fig = go.Figure()

# Add scatter plot points
fig.add_trace(go.Scatter(
    x=risk_return_df['Risk'],
    y=risk_return_df['Average Daily Return'],
    mode='markers+text',
    text=risk_return_df.index,
    textposition="top center",
    marker=dict(size=10)
))

# Update layout
fig.update_layout(
    title='Risk vs. Return Analysis',
    xaxis_title='Risk (Standard Deviation)',
    yaxis_title='Average Daily Return',
    showlegend=False
)

# Show the plot
fig.show()

**Comparative Analysis of Risk vs. Return:**

The average daily return and standard deviation (risk) for each stock are as follows:

- **Adani Enterprises (ADANIENT.NS)**:
  - Average Daily Return: -0.000604
  - Risk (Standard Deviation): 0.020704

- **ITC Ltd (ITC.NS)**:
  - Average Daily Return: 0.000930
  - Risk (Standard Deviation): 0.012271

- **Reliance Industries (RELIANCE.NS)**:
  - Average Daily Return: -0.000494
  - Risk (Standard Deviation): 0.011411

- **Sun Pharmaceutical Industries (SUNPHARMA.NS)**:
  - Average Daily Return: -0.000119
  - Risk (Standard Deviation): 0.012570

**Observations:**
- **Return Analysis**:
  - ITC Ltd (ITC.NS) has the highest average daily return among the four stocks, indicating better performance in generating returns on a daily basis.
  - Adani Enterprises (ADANIENT.NS) and Reliance Industries (RELIANCE.NS) have negative average daily returns, suggesting a slight decline in value on average per day.
  - Sun Pharmaceutical Industries (SUNPHARMA.NS) has the lowest negative average daily return, indicating the smallest decline in value per day among the four stocks.

- **Risk Analysis**:
  - Adani Enterprises (ADANIENT.NS) has the highest risk (standard deviation), indicating higher volatility in its daily returns.
  - Sun Pharmaceutical Industries (SUNPHARMA.NS) has the highest risk among the four stocks, indicating higher volatility in its daily returns.
  - Reliance Industries (RELIANCE.NS) has the lowest risk among the four stocks, suggesting relatively lower volatility in its daily returns.

**Summary**:
ITC Ltd (ITC.NS) demonstrates the highest average daily return with moderate risk. Adani Enterprises (ADANIENT.NS) and Reliance Industries (RELIANCE.NS) exhibit negative average daily returns with higher risk levels, while Sun Pharmaceutical Industries (SUNPHARMA.NS) shows a slightly lower negative return with higher risk compared to the other stocks.
